server/src/main/java/org/candlepin/controller/ActualRefresher.java

```diff
@@ -0,0 +1,1297 @@
+                    }
+
+                    updatedEntity.setBranding(branding);
+                }
+            }
+        }
+
+        // Perform child resolution
+        ProductInfo sourceEntity = (importedEntity != null ? importedEntity : existingEntity);
+
+        // Update products
+        Collection<? extends ProductInfo> providedProducts = sourceEntity.getProvidedProducts();
+
+        if (providedProducts != null) {
+            updatedEntity.setProvidedProducts(null);
+
+            for (ProductInfo product : providedProducts) {
+                EntityNode childNode = mapper.getNode(EntityNode.Type.PRODUCT, product.getId());
+
+                if (childNode == null) {
+                    String errmsg = String.format("Product references a product which does not exist: %s",
+                        product.getId());
+
+                    throw new IllegalStateException(errmsg);
+                }
+
+                if (!childNode.visited()) {
+                    String errmsg = String.format("Child node accessed before it has been processed: " +
+                        "%s => %s", sourceEntity, product);
+
+                    throw new IllegalStateException(errmsg);
+                }
+
+                updatedEntity.addProvidedProduct((Product) (childNode.changed() ?
+                    childNode.getUpdatedEntity() :
+                    childNode.getExistingEntity()));
+            }
+        }
+
+        // Update content
+        Collection<? extends ProductContentInfo> productContent = sourceEntity.getProductContent();
+
+        if (productContent != null) {
+            updatedEntity.setProductContent(null);
+
+            for (ProductContentInfo pc : productContent) {
+                ContentInfo content = pc.getContent();
+                EntityNode childNode = mapper.getNode(EntityNode.Type.CONTENT, content.getId());
+
+                if (childNode == null) {
+                    String errmsg = String.format("Product references a content which does not exist: %s",
+                        content.getId());
+
+                    throw new IllegalStateException(errmsg);
+                }
+
+                if (!childNode.visited()) {
+                    String errmsg = String.format("Child node accessed before it has been processed: " +
+                        "%s => %s", sourceEntity, content);
+
+                    throw new IllegalStateException(errmsg);
+                }
+
+                updatedEntity.addContent((Content) (childNode.changed() ?
+                    childNode.getUpdatedEntity() :
+                    childNode.getExistingEntity()),
+                    pc.isEnabled());
+            }
+        }
+
+
+        // Do version resolution
+        int version = updatedEntity.getEntityVersion();
+
+        Set<Product> candidates = vMap.<Product>getCandidateEntities(EntityNode.Type.PRODUCT,
+            updatedEntity.getId());
+
+        for (Product candidate : candidates) {
```

Comment:
`candidates` is going to need a null check before we iterate

Reply:
Yes